packages/tracing/src/hubextensions.ts

```diff
@@ -20,34 +29,150 @@ function traceHeaders(this: Hub): { [key: string]: string } {
-    // if true = we want to have the transaction
-    // if false = we don't want to have it
-    // Math.random (inclusive of 0, but not 1)
+    // we would have bailed at the beginning if neither `tracesSampler` nor `tracesSampleRate` were defined, so one of
+    // these should work; prefer the hook if so
+    const sampleRate =
+      typeof options.tracesSampler === 'function' ? options.tracesSampler(sampleContext) : options.tracesSampleRate;
+
+    // since this is coming from the user, who knows what we might get
+    if (!isValidSampleRate(sampleRate)) {
+      logger.warn(`[Tracing] Discarding trace because of invalid sample rate.`);
+      transaction.sampled = false;
+      return transaction;
+    }
+
+    // if the function returned 0, or if the sample rate is set to 0, it's a sign the transaction should be dropped
+    if (!sampleRate) {
+      logger.log(
+        `[Tracing] Discarding trace because ${
+          typeof options.tracesSampler === 'function' ? 'tracesSampler returned 0' : 'tracesSampleRate is set to 0'
+        }`,
+      );
+      transaction.sampled = false;
+      return transaction;
+    }
+
+    // now we roll the dice (Math.random is inclusive of 0, but not of 1, so strict < is safe here)
     transaction.sampled = Math.random() < sampleRate;
-  }
 
-  // We only want to create a span list if we sampled the transaction
-  // If sampled == false, we will discard the span anyway, so we can save memory by not storing child spans
-  if (transaction.sampled) {
-    const experimentsOptions = (client && client.getOptions()._experiments) || {};
-    transaction.initSpanRecorder(experimentsOptions.maxSpans as number);
+    // if we're not going to keep it, we're done
+    if (!transaction.sampled) {
+      logger.log(
+        `[Tracing] Discarding trace because it's not included in the random sample (sampling rate = ${sampleRate})`,
+      );
+      return transaction;
+    }
   }
 
+  // at this point we know we're keeping the transaction, whether because of an inherited decision or because it got
+  // lucky with the dice roll
+  const experimentsOptions = options._experiments || {};
+  transaction.initSpanRecorder(experimentsOptions.maxSpans as number);
+
   return transaction;
 }
+/**
+ * Gets the correct context to pass to the tracesSampler, based on the environment (i.e., which SDK is being used)
+ *
+ * @returns A SampleContext object
+ */
+function getDefaultSampleContext(): SampleContext {
+  const defaultSampleContext: SampleContext = {};
+
+  if (isNodeEnv()) {
+    const domain = getActiveDomain();
+
+    if (domain) {
+      // for all node servers that we currently support, we store the incoming request object (which is an instance of
+      // http.IncomingMessage) on the domain
+
+      // the domain members are stored as an array, so our only way to find the request is to iterate through the array
+      // and compare types
```

Comment:
It is in fact the existing behavior, and my understanding is that scope/thread/concurrency safety the reason we still use the deprecated `domain` package, because we don't have another way to accomplish that. @kamilogorek is the expert here, though.